### PR TITLE
Producer consumer refactor

### PR DESCRIPTION
**Pull request by joelkoz**
This pull request involves a refactor that breaks the framework down into finer grained function classes. In particular, the ValueProducer<T> and ValueConsumer<T> classes abstract out functionality from Device and Transform and let these functions stand on their own. This has been used initially in a separate project of mine to create an LCD Gauge that is a ValueConsumer<float>, which is a completely different kind of animal than a Transform.

There is one possible breaking change: ValueConsumer<T> defines set_input() as taking TWO parameters: a value, and an index. This standardizes the interface used by multi-value input consumers like Difference. Though this parameter has a default index, code that IMPLEMENTS set_input() may have to add the additional parameter to get their code to compile.

Also part of this request is some syntactic sugar: ValueProducer<T> now has a method named "connectTo(ValueConsumer<T>)" that performs the wiring performed by SensESPApp::connect_1to1(). This allows the main app to use a syntax that looks like this:

pSensor->connectTo(pTransform);

A UML diagram was added to the README.md file that diagrams the new class structure

### CODE REVIEW

**Comment by mairas** [comment]
I've now skimmed through the changes a couple of times. That's been quite a bit of effort, I have to say! Thanks a bunch! I really love what you've done here: the producer-consumer approach really clears things up and solidifies the base of the code.

I think the overall structure of the PR is good as-is. I noticed some really minor things here and there which I'd like to comment on. It's getting late, though, so if it's OK, I'll try to write down the specific comments tomorrow.

**Comment by mairas** [comment]
Finally, getting around to commenting on the PR!

As an overall style convention comment, I've tried to stick to [Google's C++ Style Guide](https://google.github.io/styleguide/cppguide.html). It does have some peculiarities, and I don't want to insist on it if you have some other style you prefer. However, I'd like to have _some_ predefined explicit style guide to follow because that makes defining a consistent style so much more easier.

Also, this PR is already quite large, so I'd rather merge it with the style inconsistencies and then work on the issue in the future.

File: src/sensesp_app.h [hunk 1]
```diff
@@ -16,6 +19,13 @@ class SensESPApp {
   void reset();
   String get_hostname();
 
+
+  template<typename T>
```
Comment (mairas) [inline src/sensesp_app.h]: I think this is much cleaner and easier to use than the previous `connect_1to1`. Do you see any reason to keep it around any longer? In any case, probably better to do the removal in a separate PR.

File: src/system/valueproducer.h [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        /**
+         * Connects this producer to the specified consumer, registering that
+         * consumer for notifications to when this producer's value changes
+         * @param inputChannel Consumers can have one or more inputs feeding them
+         *   This parameter allows you to specify which input number this producer
+         *   is connecting to. For single input consumers, leave the index at
+         *   zero.
+         *  @see ValueConsumer::set_input()
+         */
+        void connectTo(ValueConsumer<T>* pConsumer, uint8_t inputChannel = 0) {
```
Comment (mairas) [inline src/system/valueproducer.h]: Considering only the interface design, `ValueProducer::connectTo` feels like it should just be replaced everywhere with `ValueConsumer::connectFrom` because the former mostly deals with arguments referring to the latter. However, from the interface usability point of view, I think it's much easier to think in the direction of data flow and the interface is still well justified. :-)

Comment (mairas) [inline src/system/valueproducer.h]: What would it take to allow chaining of `connectTo` calls?

```
device->connectTo(transform1)->connectTo(transform2);
```

I realize it probably doesn't work for producers and consumers but it would be quite nice, nevertheless.

Comment (mairas) [reply src/system/valueproducer.h]: Ah, the next method... :-D

File: src/system/wantsenable.h [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#include <stdint.h>
+
+/**
+ * Classes that "WantsEnable" will have their enable() method
+ * call automatically at startup when the SensESP app itself
+ * is enabled.  The optional priority allows for certain classes
+ * to be initialized first. A lower priority (e.g. zero) will be
+ * enabled before a higher one (e.g. 10). The maximum priority is 10.
+ */
+class WantsEnable {
```
Comment (mairas) [inline src/system/wantsenable.h]: I'm not terribly fond of this class name. It sticks out and isn't exactly obvious: children already have an `enable` method, so why would the still want that?

How about a tongue-twister of `Enableable`?

File: src/system/wantsenable.h [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    public:
+        WantsEnable(uint8_t priority = 10);
```
Comment (mairas) [inline src/system/wantsenable.h]: This really is a minor thing, but instead of the arbitrary fixed priority range, how about using a scheme similar to Unix process priority: 0 is the default, a lower value indicates increasingly higher priority and higher (positive) values a lower one. You could then use a `std::priority_queue` for `enableList` for one-pass enabling of objects.

File: src/transforms/transform.h [hunk 5]
```diff
@@ -8,70 +8,79 @@
...
+typedef Transform<int> IntegerTransform;
+typedef Transform<String> StringTransform;
+
+
+/**
+ * A OneToOneTransform is a common type of transform that consumes, transforms,
+ * then outputs values of the same data type.
+ */
+template <typename T>
+class OneToOneTransform : public ValueConsumer<T>, public Transform<T> {
```
Comment (mairas) [inline src/transforms/transform.h]: Would you happen to have any ideas for a more descriptive name for this? If not, then this is it, I guess.

**Comment by mairas** [comment]
Finished my comment round. All in all, pretty awesome improvements!

File: src/sensesp_app.h [hunk 6]
```diff
@@ -26,10 +36,10 @@ class SensESPApp {
   template<typename T, typename U>
   void connect_2to1(T* obs1, T* obs2, U* transform) {
```
Comment (mairas) [inline src/sensesp_app.h]: I had been thinking of generalizing `connect_2to1` to take an arbitrary amount of observables/consumers. I suppose this could be doable with your new interface approach as well?

And how about multiple inputs of different types? The current codebase doesn't support that well either but I think such transforms would be quite conceivable.

**Comment by joelkoz** [comment]
> Finally, getting around to commenting on the PR!
> 
> As an overall style convention comment, I've tried to stick to [Google's C++ Style Guide](https://google.github.io/styleguide/cppguide.html). It does have some peculiarities, and I don't want to insist on it if you have some other style you prefer. However, I'd like to have _some_ predefined explicit style guide to follow because that makes defining a consistent style so much more easier.
> 
> Also, this PR is already quite large, so I'd rather merge it with the style inconsistencies and then work on the issue in the future.

I have no problem following any kind of style guide. This frequently comes down to personal preference, though it can be religious for some.  I've always felt that the original author of a project gets to pick the style, so you win! ;)  Do you know of any "pretty print" style formatting tools we can send the code through? Or, maybe a utility that can at least issue warnings in PlatformIO?  I do agree that it will be easier to merge this PR, then separately go back and rework

File: src/sensesp_app.h (hunk 6 above)
Comment (joelkoz) [reply src/sensesp_app.h]: Actually, my AnalogGauge from my reference project is both a NumericConsumer as well as a BooleanConsumer (the numeric inputs are for the possible display values, and the boolean is for a button that scrolls the display thru the various values).  Due to operator overloading, it is possible for any one consumer to be of one or more type - you end up with set_input(float, int) and set_input(bool, int).  The real trick is two inputs of the same type, but different meaning.  That is why I introduced the "input channel" concept.  What I haven't really addressed is a way for a consumer to advertise if it is capable of handling more than one input or not.  I can say I had a bug in my original AnalogGauge when I first hooked up more than one input to it.  One approach is to have a consumer discard any inputs its not capable of understanding (i.e. implement set_input() with 

if (inputChannel == 0) { process the input }

That works, but I fear that if consumers are implemented without it, it opens up the possibility for bugs.

As far as having connect_2to1 handle a variable list : it would be easy I think to have a connect_Manyto1() that auto assigned the input channel.  But, as I've worked with the code, I've found I prefer to use the new "connect" methods in the producer or consumer, vs these here in the sensesp_app.  Maybe what we really want is:

   ValueConsumer::connectManyFrom(pProducer1, pProducer2, ...)

Or better yet - make connectFrom() handle the variable parameter list directly. Then you could have

pConsumer->connectFrom(pProducer1, pProducer2, pProducer3)

File: src/system/valueproducer.h (hunk 2 above)
Comment (joelkoz) [reply src/system/valueproducer.h]: This was an artifact from the evolution of the code.  Originally, the inputChannel value was a member of the Producer, and a Producer could only connect to one and only one input channel despite the fact it could connect to multiple consumers.  When that was the case, it was impossible to implement ValueConsumer::connectFrom() (the input channel was unknown).  Once I moved the inputChannel to a parameter, I could implement connectFrom(), which I did.  But, by that time, my thinking had already shifted to the chainable call:

pDevice->connectTo(pTransform)-> connectTo(pFinalConsumer);

Now that you mention it, this chainable connectTo() version could also be implemented in ValueConsumer, though I this feels kind of awkward:

pFinalConsumer->connectFrom(pTransform)->connectFrom(pDevice)

I vote we implement them both, and let the users of the library decide which approach feels more natural for their particular use case.

Comment (joelkoz) [reply src/system/valueproducer.h]: Great minds think alike :)

File: src/system/wantsenable.h (hunk 3 above)
Comment (joelkoz) [reply src/system/wantsenable.h]: My original name was "MustEnable".  But then I though "that is not quite correct - it CAN be enabled, but its not a requirement.

I am ok with Enableable, though I will say - I just had to type that twice before I got the spelling correct. I don't feel strongly one way or another, so sure - we can change it.

Just a thought - there is nothing wrong with naming the class just a simple Enable

File: src/system/wantsenable.h (hunk 4 above)
Comment (joelkoz) [reply src/system/wantsenable.h]: I was unaware of std:priority_queue.  I like that much better.  Having a sorted list is much better. I was just being lazy, thinking the real world use of this would not warrant the extra processing needed to sort the list as it was being built (at least, that is how I justified not writing it :) )

In reality, I was unsure if this priority was even necessary.  I just knew that the previous implementation enabled the devices followed by the transforms, and when I combined them, I didn't want to break any code that might depend on this behavior.

File: src/transforms/transform.h (hunk 5 above)
Comment (joelkoz) [reply src/transforms/transform.h]: Yeah - it does feel awkward.  SymetricTransform maybe?

**Comment by mairas** [comment]
> I have no problem following any kind of style guide. This frequently comes down to personal preference, though it can be religious for some. I've always felt that the original author of a project gets to pick the style, so you win! ;) Do you know of any "pretty print" style formatting tools we can send the code through? Or, maybe a utility that can at least issue warnings in PlatformIO? I do agree that it will be easier to merge this PR, then separately go back and rework

I studied the pretty-printers a bit and `clang-format` would seem to do the job. If you create a `.clang-format` file in the project root directory with the following contents:

```
---
BasedOnStyle: Google
```

you'll get automatic formatting in Visual Studio Code. The on-the-go indentation didn't quite cope with some of the peculiarities in the Google Style (single-space indentation of public/protected/private keywords, in particular) but a "Format Selection" or "Format Document" could do those, too. But obviously no code formatting tool would be able to rename the variables etc. Let's not be too fussy about those, though.

**Comment by mairas** [comment]
And obviously we should add the `.clang-format` file in the project. :-)

File: src/system/valueproducer.h (hunk 2 above)
Comment (mairas) [reply src/system/valueproducer.h]: I agree that the chainable connectFrom is quite alien. And yes, let's implement both, but only in a later PR. Better get this merged asap.

File: src/system/wantsenable.h (hunk 3 above)
Comment (mairas) [reply src/system/wantsenable.h]: `Enable` sounds good!

File: src/system/wantsenable.h (hunk 4 above)
Comment (mairas) [reply src/system/wantsenable.h]: I think having the priority is a good idea, although it could always default to zero. I do agree that this doesn't really have a performance impact in the real life, I just prefer having as few arbitrary limits as possible.

File: src/transforms/transform.h (hunk 5 above)
Comment (mairas) [reply src/transforms/transform.h]: Not sure about SymmetricTransform - it's not really obvious, either. Let's keep the original name until something better comes along.

Comment (joelkoz) [reply src/transforms/transform.h]: I actually thought SymmeticTransform wasn't too bad.  Looking at synonyms, other candidates would be:

UniformTransform
BalancedTransform

If neither of those sound any better, then I agree - let's keep it as is for now.  Maybe the community can come up with a better name.

**Comment by joelkoz** [comment]
> Finished my comment round. All in all, pretty awesome improvements!

Thanks.  I'll got make those few changes we know we want in this PR.  Do you want to do the actual merge? Or, do you want me to merge once I check them in?

File: src/system/valueproducer.h (hunk 2 above)
Comment (joelkoz) [reply src/system/valueproducer.h]: Just had a thought: if we put a chainable version of connectFrom() in ValueConsumer(), it should probably still return "this" as a ValueProducer.  When we combine that with the ability of connectFrom() to handle a variable number or parameters, we'd end up with a syntax like this that would NOT be alien:

`pDifference->connectFrom(pInput1, pInput2) -> connectTo(pTransform2) -> pConnectTo(pFinalConsumer);`

THAT to me seems very natural.

Comment (mairas) [reply src/system/valueproducer.h]: That looks good to me!

File: src/transforms/transform.h (hunk 5 above)
Comment (mairas) [reply src/transforms/transform.h]: If you prefer SymmetricTransform over OneToOne, I'm fine with it.

File: src/sensesp_app.h (hunk 6 above)
Comment (mairas) [reply src/sensesp_app.h]: That looks good but after two seconds it's not obvious to me how correct typing would be dealt with. But it's not really a topic for this PR and can be experimented with. :-)

**Comment by mairas** [comment]
> Thanks. I'll got make those few changes we know we want in this PR. Do you want to do the actual merge? Or, do you want me to merge once I check them in?

Feel free to do the merge, it seems we're pretty much on the same page!

File: examples/rpm_counter.cpp [hunk 7]
```diff
@@ -28,10 +28,33 @@ ReactESP app([] () {
...
+/*  
+   sensesp_app->connect<int>(
+        new DigitalInputCounter(D5, INPUT_PULLUP, RISING, 500), 
+        new Frequency("propulsion.left.revolutions", 1./97., "/sensors/engine_rpm")
+   );
+*/
+
+
+// 3. Connect the producer directly to the consumer
+new DigitalInputCounter(D5, INPUT_PULLUP, RISING, 500)
```
Comment (mairas) [inline examples/rpm_counter.cpp]: Before you merge: this line seems to require parens around it. `new` probably has a lower precendence than `->`.
